src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/AvoidPotentiallyExpensiveCallWhenLoggingTests.cs

```diff
@@ -0,0 +1,3532 @@
+                            logger.Log{{logLevel}}(eventId, exception, ExpensiveMethodCall());
+                    }
+
+                    string ExpensiveMethodCall()
+                    {
+                        return "very expensive call";
+                    }
+                }
+                """;
+
+            await VerifyCSharpCodeFixAsync(source, source);
+        }
+
+        [Theory]
+        [MemberData(nameof(LogLevels))]
+        public async Task CustomLoggerGuardedWorkInLoggerMessage_NoDiagnostic_CS(string logLevel)
+        {
+            string source = $$"""
+                using System;
+                using Microsoft.Extensions.Logging;
+
+                class CustomLogger : ILogger
+                {
+                    public IDisposable BeginScope<TState>(TState state) { return default; }
+                    public bool IsEnabled(LogLevel logLevel) { return true; }
+                    public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter) { }
+                }
+
+                static partial class C
+                {
+                    [LoggerMessage(EventId = 0, Level = LogLevel.{{logLevel}}, Message = "Static log level `{message}`")]
+                    static partial void StaticLogLevel(this ILogger logger, string message);
+
+                    [LoggerMessage(EventId = 1, Message = "Dynamic log level `{message}`")]
+                    static partial void DynamicLogLevel(this ILogger logger, LogLevel level, string message);
+
+                    static void M(CustomLogger logger)
+                    {
+                        if (logger.IsEnabled(LogLevel.{{logLevel}}))
+                        {
+                            logger.StaticLogLevel(ExpensiveMethodCall());
+                        }
+
+                        if (logger.IsEnabled(LogLevel.{{logLevel}}))
+                        {
+                            logger.DynamicLogLevel(LogLevel.{{logLevel}}, ExpensiveMethodCall());
+                        }
+                    }
+
+                    static string ExpensiveMethodCall()
+                    {
+                        return "very expensive call";
+                    }
+                }
+                """;
+
+            await VerifyCSharpCodeFixAsync(source, source);
+        }
+
+        [Fact]
+        public async Task WrongLogLevelGuardedWorkInLog_ReportsDiagnostic_CS()
+        {
+            string source = """
+                using System;
+                using Microsoft.Extensions.Logging;
+
+                class C
+                {
+                    void M(ILogger logger, EventId eventId, Exception exception, Func<string, Exception, string> formatter)
+                    {
+                        if (logger.IsEnabled(LogLevel.Critical))
+                            logger.Log(LogLevel.Trace, eventId, [|ExpensiveMethodCall()|], exception, formatter);
```

Comment:
I was wondering if we should report another diagnostic if the log level does not match, as this is most likely a bug.